### PR TITLE
Effects

### PR DESCRIPTION
**Pull request by swilly22**
Support for GRAPH.EFFECT replicating modifications to both replicas and AOF

### CODE REVIEW

File: src/graph/graph_hub.c [hunk 1]
```diff
@@ -171,27 +184,38 @@ uint DeleteNode
...
 	// add edge deletion operation to undo log
-	bool      has_indecise =  GraphContext_HasIndices(gc);
-	uint      count        = array_len(edges);
-	QueryCtx *query_ctx    = QueryCtx_GetQueryCtx();
-	for (uint i = 0; i < count; i++) {
-		UndoLog_DeleteEdge(&query_ctx->undo_log, edges + i);
+	bool has_indecise = GraphContext_HasIndices(gc);
+	UndoLog *undo_log = (log == true) ? QueryCtx_GetUndoLog() : NULL;
+
+	for (uint i = 0; i < n; i++) {
```
Comment (AviAvni) [inline src/graph/graph_hub.c:206]: maybe
```
if(log == true) {
	for (uint i = 0; i < n; i++) {
		// add each deleted edge to undo-log
 		UndoLog_DeleteEdge(undo_log, edges + i);
 	}
}

if(has_indecise == true) {
	for (uint i = 0; i < n; i++) {
 		// remove edge from relevant indices
 		_DeleteEdgeFromIndices(gc, edges + i);
 	}
 }
```

**Review by AviAvni — CHANGES_REQUESTED** [review]
nice

**Comment by AviAvni** [comment]
we need to add summary of the changes and then update matrix size so the matrix policy will be not or only resize when applying effect

File: src/graph/graph_hub.c (hunk 1 above)
Comment (swilly22) [reply src/graph/graph_hub.c:206]: I don't want to perform 2N operations

File: docs/docs/configuration.md [hunk 2]
```diff
@@ -306,3 +307,22 @@ Retrieve all paths in a graph with a timeout of 500 milliseconds.
...
+`EFFECTS_THRESHOLD` is 300 μs.
+
+#### Example
+
+Assume `MATCH (n) WHERE n.id < 100 SET n.v = n.v + 1` updated 5 nodes
+and the query total execution time is 5ms, the average modification time is:
+total execution time / number of changes:  5ms / 5 = 1ms.
+if the average modification time is greater then `EFFECTS_THRESHOLD` the query
+will be replicated to both replicas and AOF as a graph effect otherwise the original
+query will be replicated.
```
Comment (LiorKogan) [inline docs/docs/configuration.md:328]: Suggestion: Add:
0: always replicate the effect; -1:  always replicate the original query.

Comment (LiorKogan) [reply docs/docs/configuration.md:328]: Please add:
(Since RedisGraph v2.12)

**Pull request merged**